lib/web_ui/lib/src/engine/profiler.dart

```diff
@@ -107,3 +107,108 @@ class Profiler {
+/// [_kFrameTimingsSubmitInterval]).
+List<ui.FrameTiming> _frameTimings = <ui.FrameTiming>[];
+
+/// The amount of time in microseconds we wait between submitting
+/// frame timings.
+const int _kFrameTimingsSubmitInterval = 100000; // 100 milliseconds
+
+/// The last time (in microseconds) we submitted frame timings.
+int _frameTimingsLastSubmitTime = _nowMicros();
+
+// These variables store individual [ui.FrameTiming] properties.
+int _vsyncStartMicros = -1;
+int _buildStartMicros = -1;
+int _buildFinishMicros = -1;
+int _rasterStartMicros = -1;
+int _rasterFinishMicros = -1;
+
+/// Records the vsync timestamp for this frame.
+void _frameTimingsOnVsync() {
+  if (!_frameTimingsEnabled) {
+    return;
+  }
+  _vsyncStartMicros = _nowMicros();
+}
+
+/// Records the time when the framework started building the frame.
+void _frameTimingsOnBuildStart() {
+  if (!_frameTimingsEnabled) {
+    return;
+  }
+  _buildStartMicros = _nowMicros();
+}
+
+/// Records the time when the framework finished building the frame.
+void _frameTimingsOnBuildFinish() {
+  if (!_frameTimingsEnabled) {
+    return;
+  }
+  _buildFinishMicros = _nowMicros();
+}
+
+/// Records the time when the framework started rasterizing the frame.
+///
+/// On the web, this value is almost always the same as [_buildFinishMicros]
+/// because it's single-threaded so there's no delay between building
+/// and rasterization.
+///
+/// This also means different things between HTML and CanvasKit renderers.
+///
+/// In HTML "rasterization" only captures DOM updates, but not work that
+/// the browser performs after the DOM updates are committed. The browser
+/// does not report that information.
+///
+/// CanvasKit captures everything because we control the rasterization
+/// process, so we know exactly when rasterization starts and ends.
+void _frameTimingsOnRasterStart() {
+  if (!_frameTimingsEnabled) {
+    return;
+  }
+  _rasterStartMicros = _nowMicros();
+}
+
+/// Records the time when the framework started rasterizing the frame.
+///
+/// See [_frameTimingsOnRasterStart] for more details on what rasterization
+/// timings mean on the web.
+void _frameTimingsOnRasterFinish() {
+  if (!_frameTimingsEnabled) {
+    return;
+  }
+  final int now = _nowMicros();
+  _rasterFinishMicros = now;
+  _frameTimings.add(ui.FrameTiming(
+    vsyncStart: _vsyncStartMicros,
+    buildStart: _buildStartMicros,
+    buildFinish: _buildFinishMicros,
+    rasterStart: _rasterStartMicros,
+    rasterFinish: _rasterFinishMicros,
+  ));
+  _vsyncStartMicros = -1;
+  _buildStartMicros = -1;
+  _buildFinishMicros = -1;
+  _rasterStartMicros = -1;
+  _rasterFinishMicros = -1;
+  if (now - _frameTimingsLastSubmitTime > _kFrameTimingsSubmitInterval) {
+    _frameTimingsLastSubmitTime = now;
+    window.invokeOnReportTimings(_frameTimings);
+    _frameTimings = <ui.FrameTiming>[];
+  }
+}
+
+/// Current timestamp in microseconds taken from the high-precision
```

Comment:
I don't know how we are going to collect these values once we are running the tests. I remember we mentioned in the future we want to run them in Firefox as well. I found this comment for firefox in the [mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now). We might want to keep it somewhere: "Firefox rounds the returned time to 1 millisecond increments."

Reply:
In the "Reduced time precision" section they explain how you can enable high precision timers. Worst case we just ignore data from Firefox if it's not useful.

Reply:
But good catch!